packages/squiggle-lang/src/rescript/GenericDist/GenericDist_GenericOperation.res

```diff
@@ -0,0 +1,156 @@
+type operation = GenericDist_Types.Operation.genericFunctionCall
+type genericDist = GenericDist_Types.genericDist
+type error = GenericDist_Types.error
+
+// TODO: It could be great to use a cache for some calculations (basically, do memoization). Also, better analytics/tracking could go a long way.
+
+type params = {
+  sampleCount: int,
+  xyPointLength: int,
+}
+
+type outputType = [
+  | #Dist(genericDist)
+  | #Error(error)
+  | #Float(float)
+  | #String(string)
+]
+
+module Output = {
+  let toDist = (o: outputType) =>
+    switch o {
+    | #Dist(d) => Some(d)
+    | _ => None
+    }
+
+  let toFloat = (o: outputType) =>
+    switch o {
+    | #Float(d) => Some(d)
+    | _ => None
+    }
+
+  let toString = (o: outputType) =>
+    switch o {
+    | #String(d) => Some(d)
+    | _ => None
+    }
+
+  let toError = (o: outputType) =>
+    switch o {
+    | #Error(d) => Some(d)
+    | _ => None
+    }
+}
+
+let fromResult = (r: result<outputType, error>): outputType =>
+  switch r {
+  | Ok(o) => o
+  | Error(e) => #Error(e)
+  }
+
+let outputToDistResult = (b: outputType): result<genericDist, error> =>
+  switch b {
+  | #Dist(r) => Ok(r)
+  | #Error(r) => Error(r)
+  | _ => Error(ImpossiblePath)
+  }
+
+let rec run = (extra, fnName: operation): outputType => {
+  let {sampleCount, xyPointLength} = extra
+
+  let reCall = (~extra=extra, ~fnName=fnName, ()) => {
+    run(extra, fnName)
+  }
+
+  let toPointSet = r => {
+    switch reCall(~fnName=#fromDist(#toDist(#toPointSet), r), ()) {
+    | #Dist(#PointSet(p)) => Ok(p)
+    | #Error(r) => Error(r)
+    | _ => Error(ImpossiblePath)
+    }
+  }
+
+  let toSampleSet = r => {
+    switch reCall(~fnName=#fromDist(#toDist(#toSampleSet(sampleCount)), r), ()) {
+    | #Dist(#SampleSet(p)) => Ok(p)
+    | #Error(r) => Error(r)
+    | _ => Error(ImpossiblePath)
+    }
+  }
+
+  let scaleMultiply = (r, weight) =>
+    reCall(
+      ~fnName=#fromDist(#toDistCombination(#Pointwise, #Multiply, #Float(weight)), r),
+      (),
+    ) |> outputToDistResult
+
+  let pointwiseAdd = (r1, r2) =>
+    reCall(
+      ~fnName=#fromDist(#toDistCombination(#Pointwise, #Add, #Dist(r2)), r1),
+      (),
+    ) |> outputToDistResult
+
+  let fromDistFn = (subFn: GenericDist_Types.Operation.fromDist, dist: genericDist) =>
+    switch subFn {
```

Comment:
regarding the name `subFn`: 

This is "sub" like in "subcategory" or "subroutine"? why is "subfunction" the descriptive name for an item of type `Types.Operation.fromDist`? 
```rescript
  // GenericDist_Types.res, line 58
  type fromDist = [
    | #toFloat(toFloat)
    | #toDist(toDist)
    | #toDistCombination(direction, arithmeticOperation, [#Dist(genericDist) | #Float(float)])
    | #toString
  ]
```

Alternatives: 
- subFnName
- fnSubName
- fnNamePrime
- fromDist 
- fromDistFn

Reply:
Agreed it's a bad name, it's too generic. Was written quickly, will rename.

Reply:
this is still `subFn` unless the diff display isn't giving me the right commit.

Reply:
You're right, sorry